File: app/remote-ui/static/plugin-readings.js
// Live entity values use their own DOM so settings drafts and chart selection survive.
function element(tag, className, text) {
  const node = document.createElement(tag);
  if (className) node.className = className;
  if (text !== undefined) node.textContent = text;
  return node;
}

/**
 * Formats one reading as {value, unit}, rescaling by device class where a
 * raw base-unit figure is unreadable.
 *
 * A plugin publishes base units -- `data_size` in bytes, `duration` in
 * seconds -- because that is what keeps a Home Assistant sensor coherent: a
 * statistic whose unit changes from KB to MB as the number grows is a broken
 * statistic. Scaling for display therefore happens here, and the unit that
 * comes back is the one the value is actually in, which is not always the
 * one the plugin published.
 *
 * Kept in step with formatPluginReading in plugin_readings.dart, deliberately:
 * the panel and Remote Admin read the same entities and must not disagree
 * about what they say.
 */
export function formatPluginReading(reading) {
  const state = reading.state;
  const unit = reading.unit || '';
  if (state == null) return { value: 'No data', unit: '' };
  if (typeof state === 'boolean') return { value: state ? 'On' : 'Off', unit: '' };
  if (typeof state === 'number') {
    if (!Number.isFinite(state)) return { value: 'No data', unit: '' };
    const deviceClass = reading.deviceClass || '';
    if (deviceClass === 'duration') return { value: duration(state, unit), unit: '' };
    if (deviceClass === 'data_size' || deviceClass === 'data_rate') {
      const scaled = dataSize(state, unit);
      if (scaled) return scaled;
    }
    const precision = Math.max(0, Math.min(6, Math.trunc(reading.accuracyDecimals || 0)));
    if (Math.abs(state) >= 1e9) return { value: state.toExponential(precision), unit };
    const value = state.toFixed(precision);
    return { value: Number(value) === 0 ? (0).toFixed(precision) : value, unit };
  }
  return { value: state === '' ? 'Empty' : String(state), unit: '' };
}

/**
 * Bytes at a readable magnitude, to one decimal. 1024 rather than 1000 and
 * labelled KB/MB/GB, the way every file manager shows it. A rate keeps its
 * denominator -- `B/min` becomes `KB/min`, never `KB`, which would turn a
 * throughput into a total.
 */
function dataSize(state, unit) {
  const slash = unit.indexOf('/');
  const base = slash === -1 ? unit : unit.slice(0, slash);
  const per = slash === -1 ? '' : unit.slice(slash);
  if (base !== 'B') return null;

  const steps = ['B', 'KB', 'MB', 'GB', 'TB', 'PB'];
  let value = Math.abs(state);
  let step = 0;
  while (value >= 1024 && step < steps.length - 1) { value /= 1024; step++; }
  const text = step === 0 ? String(Math.round(value)) : value.toFixed(1);
  return { value: (state < 0 ? '-' : '') + text, unit: steps[step] + per };
}

/**
 * A span of time as `3d 5h 2m 9s`, largest unit first, with leading and
 * trailing zero components dropped. Zero is `0s`, not nothing.
 */
function duration(state, unit) {
  let seconds = Math.round(
    unit === 'ms' ? state / 1000 : unit === 'min' ? state * 60 : unit === 'h' ? state * 3600 : state);
  const sign = seconds < 0 ? '-' : '';
  seconds = Math.abs(seconds);
  if (seconds === 0) return '0s';

  const parts = [];
  for (const [size, suffix] of [[86400, 'd'], [3600, 'h'], [60, 'm'], [1, 's']]) {
    const count = Math.floor(seconds / size);
    seconds -= count * size;
    if (count === 0 && !parts.length) continue;
    parts.push(`${count}${suffix}`);
  }
  while (parts.length > 1 && parts[parts.length - 1].startsWith('0')) parts.pop();
  return sign + parts.join(' ');
}

export function updatePluginReadings(container, readings, title = 'Readings') {
  readings = Array.isArray(readings) ? readings : [];
  container.hidden = !readings.length;
  if (!readings.length) { container.replaceChildren(); return; }
  let list = container.querySelector('dl');
  if (!list) {
    list = element('dl', 'card plugin-readings-list');
    container.append(element('div', 'card-title', title), list);
  }
  const existing = new Map([...list.children].map(row => [row.dataset.key, row]));
  for (const reading of readings) {
    const key = `${reading.type}:${reading.key}`;
    let row = existing.get(key);
    if (!row) {
      row = element('div', 'plugin-reading'); row.dataset.key = key;
      row.append(element('dt'), element('dd'));
    }
    existing.delete(key);
    const formatted = formatPluginReading(reading);
    const value = formatted.value;
    // From the formatter, not the reading: a rescaled figure carries the
    // unit it was rescaled into, and the two must not drift apart.
    const unit = reading.type === 'sensor' && reading.state != null ? formatted.unit : '';
    row.classList.toggle('multiline', value.includes('\n'));
    const label = row.querySelector('dt');
    if (label.textContent !== reading.name) label.textContent = reading.name;
    const content = row.querySelector('dd');
    content.classList.toggle('muted', reading.state == null || reading.state === '');
    const signature = JSON.stringify([value, unit]);
    if (row.dataset.value !== signature) {
      content.replaceChildren(element('span', 'plugin-reading-value', value));
      if (unit) content.append(document.createTextNode(' '), element('span', 'plugin-reading-unit', unit));
      row.dataset.value = signature;
    }
    list.append(row);
  }
  for (const row of existing.values()) row.remove();
}

File: app/lib/ui/plugin_readings.dart
import '../l10n/messages.dart';
import 'package:flutter/material.dart';

import 'kit.dart';
import 'theme.dart';

/// A plugin's confirmed entity states, displayed without editing controls.
class PluginReadings extends StatelessWidget {
  const PluginReadings({super.key, required this.readings, this.title});

  final List<Map<String, Object?>> readings;
  final String? title;

  @override
  Widget build(BuildContext context) {
    if (readings.isEmpty) return const SizedBox.shrink();
    return Column(
      crossAxisAlignment: CrossAxisAlignment.stretch,
      children: [
        SectionHeading(title ?? pluginText(context, 'Readings')),
        SettingsCard(
          children: [
            for (final reading in readings)
              _ReadingRow(
                key: ValueKey('${reading['type']}:${reading['key']}'),
                reading: reading,
              ),
          ],
        ),
      ],
    );
  }
}

/// A reading rendered for a person: the number, and the unit that number is
/// actually in, which is not always the unit the plugin published.
typedef PluginReading = ({String value, String unit});

/// Formats one reading, rescaling by device class where a raw base-unit
/// figure is unreadable.
///
/// A plugin publishes base units and says what kind of quantity it is --
/// `data_size` in bytes, `duration` in seconds -- because that is what keeps
/// a Home Assistant sensor coherent: a statistic whose unit changes from KB
/// to MB as the number grows is a broken statistic. Scaling for display
/// therefore belongs here rather than in the plugin, and doing it here fixes
/// every plugin at once. This is the same split Home Assistant's own
/// frontend makes.
///
/// Anything without a device class this understands is left exactly as it
/// was, at the precision the plugin asked for.
PluginReading formatPluginReading(Map<String, Object?> reading) {
  final state = reading['state'];
  final unit = '${reading['unit'] ?? ''}';
  if (state == null) return (value: 'No data', unit: '');
  if (state is bool) return (value: state ? 'On' : 'Off', unit: '');
  if (state is num) {
    if (!state.isFinite) return (value: 'No data', unit: '');
    final deviceClass = '${reading['deviceClass'] ?? ''}';
    if (deviceClass == 'duration') {
      return (value: _duration(state, unit), unit: '');
    }
    if (deviceClass == 'data_size' || deviceClass == 'data_rate') {
      final scaled = _dataSize(state, unit);
      if (scaled != null) return scaled;
    }
    final precision = ((reading['accuracyDecimals'] as num?)?.toInt() ?? 0)
        .clamp(0, 6);
    if (state.abs() >= 1e9) {
      return (value: state.toStringAsExponential(precision), unit: unit);
    }
    final text = state.toStringAsFixed(precision);
    return (
      value: num.parse(text) == 0 ? 0.toStringAsFixed(precision) : text,
      unit: unit,
    );
  }
  return (value: state == '' ? 'Empty' : '$state', unit: '');
}

/// Bytes at a readable magnitude, to one decimal.
///
/// 1024 rather than 1000, labelled KB/MB/GB: the powers-of-two step is what
/// every file manager and download shows, and writing KiB to be correct
/// about it would leave the row less readable than the number it replaced.
///
/// A rate keeps its denominator -- `B/min` scales to `KB/min`, not to `KB`,
/// which would silently turn a throughput into a total. Whole bytes stay
/// whole: "834 B" reads better than "834.0 B".
PluginReading? _dataSize(num state, String unit) {
  final slash = unit.indexOf('/');
  final base = slash == -1 ? unit : unit.substring(0, slash);
  final per = slash == -1 ? '' : unit.substring(slash);
  if (base != 'B') return null;

  const steps = ['B', 'KB', 'MB', 'GB', 'TB', 'PB'];
  var value = state.toDouble();
  final negative = value < 0;
  value = value.abs();
  var step = 0;
  while (value >= 1024 && step < steps.length - 1) {
    value /= 1024;
    step++;
  }
  final text = step == 0
      ? value.round().toString()
      : value.toStringAsFixed(1);
  return (value: negative ? '-$text' : text, unit: '${steps[step]}$per');
}

/// A span of time as `3d 5h 2m 9s`, largest unit first.
///
/// Leading and trailing zero components are dropped, so an hour reads `1h`
/// rather than `0d 1h 0m 0s`, and a gap in the middle is kept because
/// `1h 0m 9s` and `1h 9s` say different things about the magnitude at a
/// glance. Zero is `0s`, not nothing.
String _duration(num state, String unit) {
  var seconds = state.round();
  if (unit == 'ms') seconds = (state / 1000).round();
  if (unit == 'min') seconds = (state * 60).round();
  if (unit == 'h') seconds = (state * 3600).round();
  final sign = seconds < 0 ? '-' : '';
  seconds = seconds.abs();
  if (seconds == 0) return '0s';

  final parts = <String>[];
  for (final (size, suffix) in [
    (86400, 'd'),
    (3600, 'h'),
    (60, 'm'),
    (1, 's'),
  ]) {
    final count = seconds ~/ size;
    seconds -= count * size;
    if (count == 0 && parts.isEmpty) continue;
    parts.add('$count$suffix');
  }
  while (parts.length > 1 && parts.last.startsWith('0')) {
    parts.removeLast();
  }
  return '$sign${parts.join(' ')}';
}

class _ReadingRow extends StatelessWidget {
  const _ReadingRow({super.key, required this.reading});

  final Map<String, Object?> reading;

  @override
  Widget build(BuildContext context) {
    final theme = Theme.of(context);
    final muted = theme.colorScheme.onSurfaceVariant;
    final readingStyle = theme.textTheme.bodyLarge?.copyWith(
      fontWeight: FontWeight.w400,
    );
    final formatted = formatPluginReading(reading);
    final rawValue = formatted.value;
    final state = reading['state'];
    final value =
        state == null ||
            state is bool ||
            state == '' ||
            (state is num && !state.isFinite)
        ? pluginText(context, rawValue)
        : rawValue;
    // The unit comes from the formatter, not the reading: a rescaled figure
    // carries the unit it was rescaled into, and the two must not drift.
    final unit = reading['type'] == 'sensor' && reading['state'] != null
        ? formatted.unit
        : '';
    final multiline = value.contains('\n');
    final label = Text(
      '${reading['name']}',
      style: readingStyle?.copyWith(color: muted),
    );
    final content = Text.rich(
      TextSpan(
        text: value,
        children: [
          if (unit.isNotEmpty)
            TextSpan(
              text: ' $unit',
              style: readingStyle?.copyWith(color: muted),
            ),
        ],
      ),
      // Right-aligned even when it wraps to several lines, so a list reads
      // down the same edge as every single-line value beside it. A ragged
      // left edge is the price; a value column that changes side halfway
      // down the card was the worse one.
      textAlign: TextAlign.end,
      style: readingStyle?.copyWith(
        color: reading['state'] == null || reading['state'] == ''
            ? muted
            : theme.colorScheme.onSurface,
        fontFeatures: const [FontFeature.tabularFigures()],
      ),
    );
    return MergeSemantics(
      child: Padding(
        padding: const EdgeInsets.symmetric(horizontal: Ks.inset, vertical: 16),
        child: multiline
            ? Column(
                crossAxisAlignment: CrossAxisAlignment.stretch,
                children: [label, const SizedBox(height: 8), content],
              )
            : Row(
                crossAxisAlignment: CrossAxisAlignment.start,
                children: [
                  Expanded(flex: 9, child: label),
                  const SizedBox(width: 16),
                  Expanded(flex: 11, child: content),
                ],
              ),
      ),
    );
  }
}
